eppo_core/src/attributes.rs

```diff
@@ -1,8 +1,48 @@
-use std::collections::HashMap;
+use std::{collections::HashMap, sync::Arc};
 
 use derive_more::From;
 use serde::{Deserialize, Serialize};
 
+use crate::Str;
+
+/// `Subject` is a bundle of subject attributes and a key.
+#[derive(Debug)]
+pub(crate) struct Subject {
+    /// Subject key encoded as attribute value. Known to be `AttributeValue::String`. This is
+    /// done to allow returning subject key as an attribute when rule references "id".
+    key: AttributeValue,
+    attributes: Arc<Attributes>,
+}
+
+impl Subject {
+    pub fn new(key: Str, attributes: Arc<Attributes>) -> Subject {
+        Subject {
+            key: AttributeValue::String(key),
+            attributes,
+        }
+    }
+
+    pub fn key(&self) -> &Str {
+        let AttributeValue::String(s) = &self.key else {
+            unreachable!("Subject::key is always encoded as AttributeValue::ArcString()");
+        };
+        s
+    }
+
+    pub fn get_attribute(&self, name: &str) -> Option<&AttributeValue> {
+        let value = self.attributes.get(name);
+        if value.is_some() {
+            return value;
+        }
+
+        if name == "id" {
+            return Some(&self.key);
+        }
```

Comment:
I imagine the ordering is such that when someone explicitly sets `"id"`, that takes precedence, which is great! But I think it might be nice to add that as a comment for our future selves